src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp

```diff
@@ -5545,6 +5545,14 @@ address MacroAssembler::arrays_equals(Register a1, Register a2, Register tmp3,
   int length_offset = arrayOopDesc::length_offset_in_bytes();
   int base_offset
     = arrayOopDesc::base_offset_in_bytes(elem_size == 2 ? T_CHAR : T_BYTE);
+  // When the base is not aligned to 8 bytes, then we let
+  // the compare loop include the array length, and skip
+  // the explicit comparison of length.
+  bool is_8aligned = is_aligned(base_offset, BytesPerWord);
+  assert(is_aligned(base_offset, BytesPerWord) || is_aligned(length_offset, BytesPerWord),
+         "base_offset or length_offset must be 8-byte aligned");
```

Comment:
Not that I see this changing. But the correctness of using `length_offset` relies on the length and the payload being consecutive in memory. Should probably assert this.
```suggestion
  assert(is_aligned(base_offset, BytesPerWord) || is_aligned(length_offset, BytesPerWord),
         "base_offset or length_offset must be 8-byte aligned");
  assert(is_aligned(base_offset, BytesPerWord) || base_offset == length_offset + BytesPerInt,
         "base_offset must be 8-byte aligned or no padding between base and length");
```